File: medplum-java/medplum-model/src/main/java/com/fhir/CompositionStatus.java
package com.fhir;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Locale;
import javax.annotation.Nonnull;

public final class CompositionStatus {
  public static final CompositionStatus AMENDED = new CompositionStatus(Value.AMENDED, "AMENDED");

  public static final CompositionStatus PRELIMINARY = new CompositionStatus(Value.PRELIMINARY, "PRELIMINARY");

  public static final CompositionStatus FINAL = new CompositionStatus(Value.FINAL, "FINAL");

  private final Value value;

  private final String string;

  CompositionStatus(Value value, String string) {
    this.value = value;
    this.string = string;
  }

  public Value getEnumValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return this.string;
  }

  @Override
  public boolean equals(Object other) {
    return (this == other) 
      || (other instanceof CompositionStatus && this.string.equals(((CompositionStatus) other).string));
  }

  @Override
  public int hashCode() {
    return this.string.hashCode();
  }

  public <T> T accept(Visitor<T> visitor) {
    switch (value) {
      case AMENDED:
        return visitor.visitAMENDED();
      case PRELIMINARY:
        return visitor.visitPRELIMINARY();
      case FINAL:
        return visitor.visitFINAL();
      case UNKNOWN:
      default:
        return visitor.visitUnknown(string);
    }
  }

  @JsonCreator(
      mode = JsonCreator.Mode.DELEGATING
  )
  public static CompositionStatus valueOf(@Nonnull String value) {
    String upperCasedValue = value.toUpperCase(Locale.ROOT);
    switch (upperCasedValue) {
      case "AMENDED":
        return AMENDED;
      case "PRELIMINARY":
        return PRELIMINARY;
      case "FINAL":
        return FINAL;
      default:
        return new CompositionStatus(Value.UNKNOWN, upperCasedValue);
    }
  }

  public enum Value {
    PRELIMINARY,

    FINAL,

    AMENDED,

    UNKNOWN
  }

  public interface Visitor<T> {
    T visitPRELIMINARY();

    T visitFINAL();

    T visitAMENDED();

    T visitUnknown(String unknownType);
  }
}

File: medplum-typescript/src/model/fhir/CompositionStatus.ts
export type CompositionStatus =
    | CompositionStatus.Preliminary
    | CompositionStatus.Final
    | CompositionStatus.Amended;

export const CompositionStatus = {
    Preliminary: "preliminary" as CompositionStatus.Preliminary,
    Final: "final" as CompositionStatus.Final,
    Amended: "amended" as CompositionStatus.Amended,

    _visit: <Result>(
        value: CompositionStatus,
        visitor: CompositionStatus._Visitor<Result>
    ): Result => {
        switch (value) {
            case CompositionStatus.Preliminary:
                return visitor.preliminary();
            case CompositionStatus.Final:
                return visitor.final();
            case CompositionStatus.Amended:
                return visitor.amended();
            default:
                return visitor._unknown();
        }
    },
};

export declare namespace CompositionStatus {
    export type Preliminary = "preliminary" & {
        __Preliminary: void;
    };
    export type Final = "final" & {
        __Final: void;
    };
    export type Amended = "amended" & {
        __Amended: void;
    };

    export interface _Visitor<Result> {
        preliminary: () => Result;
        final: () => Result;
        amended: () => Result;
        _unknown: () => Result;
    }
}
